### PR TITLE
STEP for transformer refactor

### PR DESCRIPTION
**Pull request by fkiraly**
This is a STEP for the transformer refactor, which will in turn rely on preparations in sktime PR 1125.

FYI @mloning, @TonyBagnall, @aiwalter, @chrisholder, @guzalbulatova, @sveameyer13

### CODE REVIEW

File: steps/11_transformer_refactor/step.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+* series-to-primitives
+* panel-to-tabular
+* series-to-series
+* panel-to-panel
+* series-to-panel
+
+Where series-to-panel occurs only in two places currently and not formalised as a transformer type:
+* sliding window forecasting (in `update_predict`)
+* epoching, in `sktime-neuro`
```
Comment (SveaMeyer13) [inline steps/11_transformer_refactor/step.md:40]: Hi @fkiraly, I'm not entirely sure whether we should handle the epoching as a transformer. (not a strong opinion just bouncing off my thoughts:) What the use case here usually requires is that we can pass a time series and one or more labels, e.g. "right" and "left" and an interval e.g. [-0.5 1]. And then this function/transformer cuts out all the parts of the data where these events occurred. It then returns panel data and a labels vector (describing each instance of the panel). So the return structure would be similar to the one you get when loading a toy dataset in sklearn (it simply returns X and y). And it could then further be used just like a normal classification problem. The critical point I see is that a transformer would not be able to return the labels vector y but would require it to be fixed in the beginning. Which I think is a bit unhandy.

I see two options of forcing this use case into a transformer: 
(1) We could make the user specify only one label. Then the label of each instance in the panel would be clear. And the user would then have to apply this transformer per label, e.g. first cutting out all the "right" trials and then cutting out all the "left" trials. Here we would however have the risk that the user (accidentally) applies different panel-to-panel transformers to the two generated panels afterwards, thereby classifying what happened to the data afterwards rather than purely classifying the label.  
 
(2) We could require the labels vector (y) (indicating which instance of the panel is supposed to be of which label afterwards) to be fixed beforehand. This would basically require a function `create_labels_vector` that creates the labels vector based on the labels we want to cut out and an annotation object (currently a `pd.DataFrame` that contains events and onsets). So instead of calling simply one function we would require the user to first create the labels vector and then to `fit_transform()` the series.

File: steps/11_transformer_refactor/step.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+We replace the old taxonomy with an input/output and rowwise yes/no taxonomy - based on what type an individual instance has, and whether `fit`/`transform` is independent per-instance or not:
+
+| `transform-input` | `transform-output` | `rowwise` | old name |
+|---|---|---|---|
+| `Series` | `Primitives` | True | series-to-primitives |
+| `Series` | `Primitives` | False | panel-to-tabular |
```
Comment (SveaMeyer13) [inline steps/11_transformer_refactor/step.md:47]:
```suggestion
| `Panel` | `Primitives` | False | panel-to-tabular |
```

File: steps/11_transformer_refactor/step.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+| `Series` | `Series` | True | series-to-series |
+| `Series` | `Series` | False | panel-to-panel |
```
Comment (SveaMeyer13) [inline steps/11_transformer_refactor/step.md:49]:
```suggestion
| `Panel` | `Panel` | False | panel-to-panel |
```

File: steps/11_transformer_refactor/step.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+Only the five combinations above are allowed (implemented), plus "primitives-to-primitives" which is the tag combination for scikit-learn type transformers.
+
+To avoid confusion, the old terminology "series-to-series" etc is no longer used, instead we talk about input/output types.
+The following table gives a translation between old and new scheme.
+
+| `transform-input` | `transform-output` | `rowwise` | old name |
+|---|---|---|---|
+| `Series` | `Primitives` | True | series-to-primitives |
+| `Series` | `Primitives` | False | panel-to-tabular |
```
Comment (SveaMeyer13) [inline steps/11_transformer_refactor/step.md:117]:
```suggestion
| `Panel` | `Primitives` | False | panel-to-tabular |
```

File: steps/11_transformer_refactor/step.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+| `Series` | `Series` | True | series-to-series |
+| `Series` | `Series` | False | panel-to-panel |
```
Comment (SveaMeyer13) [inline steps/11_transformer_refactor/step.md:119]:
```suggestion
| `Panel` | `Panel` | False | panel-to-panel |
```

File: steps/11_transformer_refactor/step.md (hunk 1 above)
Comment (fkiraly) [reply steps/11_transformer_refactor/step.md:40]: This is how I see it would be part of the framework:

* it would be a "series-to-panel" transformer which takes a `y` that is of segmentation scitype.. 
* the segments (in your diction: intervals, e.g., [-0.5, 1]) would be contained in `y`, e.g., as an `IntervalIndex` or in some other encoding. 
* the series would be passed in `X`. The return of transform is a panel which contains all the "cut out" sub-sequences.

Comment (fkiraly) [reply steps/11_transformer_refactor/step.md:40]: so sth like this

```python
mytrafo = SegmentCutter()
y = pd.IntervalIndex.from_arrays([-0.5, 10, 14], [1, 12, 15])
epochs = mytrafo.fit_transform(X, y)
```

now `epochs` is a panel which contains cut parts form `X`, in intervals [-0.5, 1], [10,12], and [14,15].

File: steps/11_transformer_refactor/step.md (hunk 2 above)
Comment (fkiraly) [reply steps/11_transformer_refactor/step.md:47]: no, no - this is right and intended like this!
The `Series` refers to "what is in one row". 
Multiple instances of `Series` make a `Panel`, and multiple instances of `Primitives` make a table (doesn't have a type).

File: steps/11_transformer_refactor/step.md (hunk 3 above)
Comment (fkiraly) [reply steps/11_transformer_refactor/step.md:49]: also this is intended the way it is

File: steps/11_transformer_refactor/step.md (hunk 2 above)
Comment (fkiraly) [reply steps/11_transformer_refactor/step.md:47]: of course, if you think it *should* be different from what I intended, please let me know (I think there are arguments for that too).
The reply is based on me thinking that you were thinking this was an unintended mistake rather than explicit intention on my side.

Comment (fkiraly) [reply steps/11_transformer_refactor/step.md:47]: come to think of it, maybe the explanation of what I was thinking could be improved, too?

File: steps/11_transformer_refactor/step.md (hunk 1 above)
Comment (SveaMeyer13) [reply steps/11_transformer_refactor/step.md:40]: But how would you know the labels of the instances in epochs? You'd have to have an additional function that creates the labels vector from y in order to train a classifier, right?

File: steps/11_transformer_refactor/step.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+We replace the old taxonomy with an input/output and rowwise yes/no taxonomy - based on what type an individual instance has, and whether `fit`/`transform` is independent per-instance or not:
+
+| `transform-input` | `transform-output` | `rowwise` | old name |
+|---|---|---|---|
+| `Series` | `Primitives` | True | series-to-primitives |
+| `Series` | `Primitives` | False | panel-to-tabular |
+| `Series` | `Series` | True | series-to-series |
+| `Series` | `Series` | False | panel-to-panel |
+| `Series` | `Panel` | True | N/A, this is new |
```
Comment (SveaMeyer13) [inline steps/11_transformer_refactor/step.md:50]: I now understand how the above was intended. But then why do you have `panel` here?

Comment (fkiraly) [reply steps/11_transformer_refactor/step.md:50]: well, because this is the only case where an individual series gets "changed" into a panel, i.e., one series input corresponds to one panel output.

In all the other cases, you have one-to-one correspondence between individual series and/or dataframe rows, on the index level.

File: steps/11_transformer_refactor/step.md (hunk 1 above)
Comment (fkiraly) [reply steps/11_transformer_refactor/step.md:40]: ah, I see - you have labels!
Well, in that case, you would start with a labelled segmentation, let's say `Y`, which is a `pd.Series` with a `pd.IntervalIndex`, and entries are labels.
You could now pass `y=Y.index` to the `SegmentCutter` as above, and `Y.reset_index()` would give you the labels - unless you allow panels which can also have a `pd.intervalIndex` as instance index, in which case `epochs` and `Y` already would have the same index and could be used in a classifier.

**Review by mloning — COMMENTED** [review]
Hi @fkiraly, I had a quick look.

I'm in favour of the refactoring along the lines of the forecasting module with public and inner fit methods. 

The proposed change in `fit(X: Panel or Series, y=None) -> self` is a big interface change. I think we need to move carefully here and see how that works on in terms of efficiency and maintainability. I'm not aware of a way of writing a general rowwise applications of a transformation without looping in Python which would be very inefficient.

**Comment by fkiraly** [comment]
>I'm not aware of a way of writing a general rowwise applications of a transformation without looping in Python which would be very inefficient.

`numpy.apply_over_axes`, `numpy.apply_along_axis`, `pandas.DataFrame.apply`, and - in case of many transformations - making used of vectorized operations in `numpy` or `pandas`.

Many algorithms can be implemented, and in many cases of the interfaced algorithms, are already implemented in vectorized fashion.

>The proposed change in fit(X: Panel or Series, y=None) -> self is a big interface change

It is, but we can make it so it is fully downwards compatible - it would just add to the series transformers the capability to run on panels. Currently, you have to iterate over rows, so the current way of doing this is very inefficient.

**Comment by mloning** [comment]
> numpy.apply_over_axes, numpy.apply_along_axis, pandas.DataFrame.apply, and - in case of many transformations - making use of vectorized operations in numpy or pandas.

As far as I'm aware, in many cases, they're not much more efficient than for loops and you don't get fitted parameters back from them when applying `fit_transform`.

**Comment by fkiraly** [comment]
>you don't get fitted parameters back from them when applying `fit_transform`.

What do you mean by that? Why would you even expect to get fitted parameters when applying `numpy.apply_etc`? Or, what are you comparing to?

**Comment by mloning** [comment]
What I'm thinking about is this: When you apply a transformer to each row via `fit_transform`, you may not only want to transform the data but also keep the fitted parameters for each row-wise application. I'm not sure how this can be achieved in an efficient way. 

With the refactoring, is it still possible to write different implementations for series-to-series and panel-to-panel transformers? 

I haven't had much more time to look at the proposal in more depth. My general feeling is that we're moving a bit too quickly at the moment, we're introducing many API changes without giving them much time to trickle through to all modules. Not sure if I'm the only one who thinks that?
